File: java/v1/src/main/java/com/api/v1/services/car/AllCarsDeletionServiceImpl.java
package com.api.v1.services.car;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.api.v1.annotations.car.VIN;
import com.api.v1.domain.car.CarRepository;
import com.api.v1.exceptions.car.EmptyCarEntityException;
import com.api.v1.utils.car.CarFinderUtil;

import reactor.core.publisher.Mono;

@Service
class AllCarsDeletionServiceImpl implements AllCarsDeletionService {

    @Autowired
    private CarRepository repository;

    @Autowired
    private CarFinderUtil carFinderUtil;

    @Override
    public Mono<Void> deleteAll() {
        return repository
                .findAll()
                .hasElements()
                .flatMap(hasElements -> {
                    if (!hasElements) return Mono.error(EmptyCarEntityException::new);
                    return repository.deleteAll();
                });
    }

    @Override
    public Mono<Void> deleteByVin(@VIN String vin) {
        return carFinderUtil
                .find(vin)
                .flatMap(car -> repository.delete(car));
    }

}

File: v2/src/main/kotlin/com/api/v2/car/services/AllCarsDeletionServiceImpl.kt
package com.api.v2.car.services

import com.api.v2.car.domain.CarRepository
import com.api.v2.car.exceptions.EmptyCarEntityException
import com.api.v2.car.utils.CarFinderUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class AllCarsDeletionServiceImpl: AllCarsDeletionService {

    @Autowired
    lateinit var carRepository: CarRepository

    @Autowired
    lateinit var carFinderUtil: CarFinderUtil

    override suspend fun deleteAll() {
        withContext(Dispatchers.IO) {
            if (checkIfCarHasNoElement()) {
                throw EmptyCarEntityException()
            }
            carRepository.deleteAll()
        }
    }

    suspend fun checkIfCarHasNoElement(): Boolean {
        return carRepository.count() == 0L
    }

    override suspend fun deleteByVin(vin: String) {
        withContext(Dispatchers.IO) {
            val existingCar = carFinderUtil.find(vin)
            carRepository.delete(existingCar)
        }
    }

}